trace/tracestate.go

```diff
@@ -0,0 +1,183 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trace
+
+import (
+	"container/list"
+	"errors"
+	"fmt"
+)
+
+const (
+	// KeyMaxSize is the maximum characters allowed in the key.
+	KeyMaxSize = 256
+
+	// ValueMaxSize is the maximum characters allowed in the value.
+	ValueMaxSize = 256
+
+	// MaxKeyValuePairs is the maximum number of key-value pairs allowed in the tracestate.
+	MaxKeyValuePairs = 32
+)
+
+// Tracestate represent tracing-system specific context in a list of key-value pairs. Tracestate allows different
+// vendors propagate additional information and inter-operate with their legacy Id formats.
+type Tracestate struct {
+	stateList *list.List
+}
+
+// Entry represent one key-value pair in a list of key-value pair of Tracestate.
+// Key is an opaque string up to 256 characters printable. It MUST begin with a lowercase letter,
+// and can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
+// forward slashes /.
+//
+// Value is an opaque string up to 256 characters printable ASCII RFC0020 characters (i.e., the
+// range 0x20 to 0x7E) except comma , and =.
+type Entry struct {
+	key   string
+	value string
+}
+
+// Get retrieves value for a given key from Tracestate ts.
+// If the key is not found then the value "" is returned.
+// If the key is found then its value is returned.
+func (ts *Tracestate) Get(key string) string {
+	if ts.stateList == nil {
+		return ""
+	}
+	var kv *Entry
+	for e := ts.stateList.Front(); e != nil; e = e.Next() {
+		kv = e.Value.(*Entry)
+		if kv.key == key {
+			return kv.value
+		}
+	}
+	return ""
+}
+
+// GetEntries retrieves all key/value Entry from Tracestate ts.
+func (ts *Tracestate) GetEntries() *list.List {
+	return ts.stateList
+}
+
+// Remove removes the key/value Entry the Tracestate ts.
+func (ts *Tracestate) Remove(key string) {
```

Comment:
Do we want to follow the List.Remove signature - return the item that has been removed?

```
 func (l *List) Remove(e *Element) interface{} {
  	if e.list == l {
  		// if e.list == l, l must have been initialized when e was inserted
  		// in l or l == nil (e is a zero Element) and l.remove will crash
  		l.remove(e)
  	}
  	return e.Value
  }
```